### DISCUSSION TITLE
Can't access ENV vars defined in .env.test

### DISCUSSION THREAD

**Discussion by erkde** [vercel/next.js#16270] [Help] [answered]
# Bug report

## Describe the bug

I have added an .env.test file to an app based on the environment variable starter, containing one variable TEST_ENV_VARIABLE.

When I run the tests, ENV_VARIABLE defined in .env is appearing in process.env, but not TEST_ENV_VARIABLE defined in .env.test

## To Reproduce

Steps to reproduce the behaviour, please provide code snippets or a repository:

1. git clone https://github.com/erkde/environment-variables-app
2. yarn install
3. yarn test

## Expected behaviour

To be able to add/override env variables using an .env.test file within tests.

## Screenshots

N/A

## System information

- OS: [e.g. macOS]
- Browser (if applies) [e.g. chrome, safari]
- Version of Next.js: [e.g. 9.5.2]
- Version of Node.js: [e.g. 12.18.2]

**Comment by jportela** [comment]
This seems to be the same issue as in https://github.com/vercel/next.js/discussions/16315

You may need to instantiate NextJS as part of Jest setup for the .env.test file to be loaded:

```js
// jest setup file
import next from 'next'
next({})
```

**Comment by jportela** [comment]
Yeah, I totally agree with you. For that reason I’ve decided to use dotenv on my test setup, instead of instantiating Next. That way I can still use a .env.test file, without the overhead of having to build first.

Documentation is definitely confusing, since it leads you to think this would work automatically in Jest, when that’s not the case. I’ve submitted a PR to clarify that.

Having the loadEnv logic exportable is interesting for sure, in my case it would avoid having to import dotenv just for my tests and use the same logic as nextjs does

**Reply by justinjaeger** [reply]
Hi, the solution by using next({}) in my jest.setup.js file works for now, but what did you mean by you "decided to use dotenv on my test setup instead of instantiating Next", because it is quite tedious to run the build every time I want to test. Thanks!

**Reply by justinjaeger** [reply]
Actually never mind, I solved it by checking out the post that erkde send #17903.

Inside jest.setup.js:

import { loadEnvConfig } from '@next/env'

export default async () => {
  loadEnvConfig(process.env.PWD)
};

Inside jest.config.js:
globalSetup: '<rootDir>/jest.setup.js'

In case that doesn't work for someone, my entire jest.config file looks like this:

module.exports = {
  verbose: true,
  modulePaths: ["<rootDir>"],
  setupFiles: ["dotenv/config"],
  testEnvironment: 'node',
  collectCoverageFrom: [
    '**/*.{js,jsx,ts,tsx}',
    '!**/*.d.ts',
    '!**/node_modules/**',
  ],
  setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
  // testPathIgnorePatterns: ['/node_modules/', '/.next/'],
  transform: {
    '^.+\\.(js|jsx|ts|tsx)$': '<rootDir>/node_modules/babel-jest'
  },
  transformIgnorePatterns: [
    '/node_modules/',
    '^.+\\.module\\.(css|sass|scss)$',
  ],
  moduleNameMapper: {
    '^.+\\.module\\.(css|sass|scss)$': 'identity-obj-proxy',
  },
  globalSetup: '<rootDir>/jest.setup.js',
}

**Reply by erkde** [reply]
You might not need `setupFiles` and `setupFilesAfterEnv`, at least in our project, loading the environment variables using `globalSetup: '/jest.setup.js'` was ok.

**Comment by qkreltms** [comment]
It seems you need to build your next application first to inject .env.test

**Reply by erkde** [reply]
@qkreltms in newer versions, you won't need to, at least the last time I was working on some NextJS code, this was working ok https://github.com/vercel/next.js/issues/17903#issuecomment-733994841

**Reply by eric-burel** [reply]
Edited my comment: it's now better documented: https://nextjs.org/docs/basic-features/environment-variables#test-environment-variables

**Comment by leerob** [comment, marked as answer]
Update: https://github.com/vercel/next.js/pull/38359
